File: selecion.java
import java.util.*;
import java.util.stream.Collectors;

public class GA_SelectionOnly {

    static final int POP_SIZE = 8;
    static final int CHROM_LENGTH = 6;
    static final int SELECT_COUNT = 4;

    static Random random = new Random();

    // ---------- CREATE RANDOM POPULATION ----------
    static List<String> create_population(int size, int length) {
        List<String> population = new ArrayList<>();
        for (int i = 0; i < size; i++) {
            StringBuilder chrom = new StringBuilder();
            for (int j = 0; j < length; j++) {
                chrom.append(random.nextBoolean() ? '1' : '0');
            }
            population.add(chrom.toString());
        }
        return population;
    }

    // ---------- FITNESS FUNCTION ----------
    static int fitness_function(String chromosome) {
        int count = 0;
        for (char c : chromosome.toCharArray()) {
            if (c == '1') count++;
        }
        return count;  // maximize number of 1's
    }

    // ---------- CANONICAL SELECTION ----------
    static List<String> canonical_selection(List<String> pop, List<Integer> fit, int k) {
        List<Integer> idx = new ArrayList<>();
        for (int i = 0; i < fit.size(); i++) idx.add(i);

        // Sort population by descending fitness
        idx.sort((a, b) -> fit.get(b) - fit.get(a));

        List<String> selected = new ArrayList<>();
        for (int i = 0; i < k; i++)
            selected.add(pop.get(idx.get(i)));

        return selected;
    }

    // ---------- ROULETTE WHEEL ----------
    static List<String> roulette_selection(List<String> pop, List<Integer> fit, int k) {
        double totalFit = fit.stream().mapToDouble(f -> f).sum();
        List<String> selected = new ArrayList<>();

        for (int s = 0; s < k; s++) {
            double r = random.nextDouble() * totalFit;
            double cum = 0;
            for (int i = 0; i < pop.size(); i++) {
                cum += fit.get(i);
                if (cum >= r) {
                    selected.add(pop.get(i));
                    break;
                }
            }
        }
        return selected;
    }

    // ---------- TOURNAMENT SELECTION ----------
    static List<String> tournament_selection(List<String> pop, List<Integer> fit, int k, int tSize) {
        List<String> selected = new ArrayList<>();

        for (int s = 0; s < k; s++) {
            int bestIdx = random.nextInt(pop.size());

            for (int i = 1; i < tSize; i++) {
                int competitor = random.nextInt(pop.size());
                if (fit.get(competitor) > fit.get(bestIdx)) {
                    bestIdx = competitor;
                }
            }
            selected.add(pop.get(bestIdx));
        }
        return selected;
    }

    // ---------- MAIN ----------
    public static void main(String[] args) {

        List<String> population = create_population(POP_SIZE, CHROM_LENGTH);
        List<Integer> fitness = population.stream().map(GA_SelectionOnly::fitness_function).collect(Collectors.toList());

        System.out.println("Initial Population:");
        for (int i = 0; i < population.size(); i++)
            System.out.println(population.get(i) + " -> Fit = " + fitness.get(i));

        System.out.println("\n--- Canonical Selection ---");
        System.out.println(canonical_selection(population, fitness, SELECT_COUNT));

        System.out.println("\n--- Roulette Wheel Selection ---");
        System.out.println(roulette_selection(population, fitness, SELECT_COUNT));

        System.out.println("\n--- Tournament Selection (size = 3) ---");
        System.out.println(tournament_selection(population, fitness, SELECT_COUNT, 3));
    }
}

File: selecion.py
import random

POP_SIZE = 8
CHROM_LENGTH = 6
SELECT_COUNT = 4


# ---------- CREATE RANDOM POPULATION ----------
def create_population(size, length):
    population = []
    for _ in range(size):
        chrom = "".join("1" if random.random() < 0.5 else "0" for _ in range(length))
        population.append(chrom)
    return population


# ---------- FITNESS FUNCTION ----------
def fitness_function(chromosome):
    return sum(1 for c in chromosome if c == '1')


# ---------- CANONICAL SELECTION ----------
def canonical_selection(pop, fit, k):
    idx = list(range(len(fit)))
    idx.sort(key=lambda i: fit[i], reverse=True)  # descending fitness

    selected = []
    for i in range(k):
        selected.append(pop[idx[i]])

    return selected


# ---------- ROULETTE WHEEL SELECTION ----------
def roulette_selection(pop, fit, k):
    total_fit = sum(fit)
    selected = []

    for _ in range(k):
        r = random.random() * total_fit
        cum = 0
        for i in range(len(pop)):
            cum += fit[i]
            if cum >= r:
                selected.append(pop[i])
                break

    return selected


# ---------- TOURNAMENT SELECTION ----------
def tournament_selection(pop, fit, k, t_size):
    selected = []

    for _ in range(k):
        best_idx = random.randint(0, len(pop) - 1)
        for _ in range(1, t_size):
            competitor = random.randint(0, len(pop) - 1)
            if fit[competitor] > fit[best_idx]:
                best_idx = competitor
        selected.append(pop[best_idx])

    return selected


# ---------- MAIN EXECUTION ----------
def main():
    population = create_population(POP_SIZE, CHROM_LENGTH)
    fitness = [fitness_function(ch) for ch in population]

    print("Initial Population:")
    for chrom, fit in zip(population, fitness):
        print(f"{chrom} -> Fit = {fit}")

    print("\n--- Canonical Selection ---")
    print(canonical_selection(population, fitness, SELECT_COUNT))

    print("\n--- Roulette Wheel Selection ---")
    print(roulette_selection(population, fitness, SELECT_COUNT))

    print("\n--- Tournament Selection (size = 3) ---")
    print(tournament_selection(population, fitness, SELECT_COUNT, 3))


if _name_ == "_main_":
    main()

#=============visualisaion=======
import random
import matplotlib.pyplot as plt

# ============================
# Genetic Algorithm Components
# ============================

POP_SIZE = 8
CHROM_LENGTH = 6

def create_population(size, length):
    population = []
    for _ in range(size):
        chrom = "".join("1" if random.random() < 0.5 else "0" for _ in range(length))
        population.append(chrom)
    return population

def fitness_function(chromosome):
    return sum(1 for c in chromosome if c == '1')


# ============================
# Visualization Function
# ============================

def visualize_fitness(population, fitness):
    plt.figure(figsize=(8, 5))
    plt.bar(range(len(fitness)), fitness)
    plt.xlabel("Individual Index")
    plt.ylabel("Fitness Value")
    plt.title("Fitness Visualization of Population")
    plt.tight_layout()
    plt.show()


# ============================
# MAIN EXECUTION
# ============================

if _name_ == "_main_":
    population = create_population(POP_SIZE, CHROM_LENGTH)
    fitness = [fitness_function(ch) for ch in population]

    print("Population:", population)
    print("Fitness:   ", fitness)

    visualize_fitness(population, fitness)
